File: medium/removekdigits.cpp
#include <iostream>

class Solution {
public:
    string removeKdigits(string num, int k) {
        stack<char> stk;

        for(char digit : num) {
            while(!stk.empty() && stk.top() > digit && k > 0) {
                stk.pop();
                k -= 1;
            }
            stk.push(digit);
        }

        while(k > 0 && !stk.empty()) {
            stk.pop();
            k -= 1;
        }

        string temp = "";
        while(!stk.empty()) {
            temp.push_back(stk.top());
            stk.pop();
        }
        reverse(temp.begin(), temp.end());

        string res = "";
        int found = 0, m = temp.size();
        for(int i = 0; i < m; i++) {
            if(temp[i] == '0' && found == 0) {
                continue;
            } else {
                res.push_back(temp[i]);
                found = 1;
            }
        }

        if(res.size() == 0) {
            res.push_back('0');
        }

        return res;
    }
};

File: python/medium/removekdigits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        sk = []

        for digit in num:
            while sk and sk[-1] > digit and k > 0:
                sk.pop()
                k -= 1
            
            sk.append(digit)

        while k > 0 and sk:
            sk.pop()
            k -= 1

        temp = ""
        n = len(sk)
        for i in range(n):
            temp += sk[-1]
            sk.pop()
            
        temp = temp[::-1]

        ret = ""
        found = 0
        m = len(temp)
        for i in range(m):
            if temp[i] is '0' and found is 0:
                continue
            else:
                ret += temp[i]
                found = 1

        if len(ret) is 0:
            ret += '0'

        return ret

